### ISSUE TITLE
Provide first-class access to environment variables?

### ISSUE THREAD

**Issue by glenjamin** [james-henderson/phoenix#3]
Hello!

I really like the look of phoenix, although I was mildly put off by the examples having all of their config in files.

From what I can tell in the examples, it would be fairly easy for me to create a component which reads the environment and is depended on by other components to get the values out - but I think it could be useful to provide support for this out-of-the-box in the http://12factor.net/config manner.

It could work something like this:

``` clj
{::env {:db-uri "user:pass@dev-db.mycompany.com"  ; specify defaults on the ::env key
        :nrepl-port 5432
        :http-port 1234}
 :database {::component myapp.database/make
            :uri [::env :db-uri]} ; similar syntax to ::deps
 :nrepl {::component myapp.nrepl/make
         :port [::env :nrepl-port]}
 :ring {::component myapp.handler/make
        :port [::env :http-port]}

```

**Comment by jarohen** [comment]
Good idea - although I'd be tempted to put the default in with the `[::env ...]`? Better to have it close to the actual declaration, rather than have to look through the config file for it.

Thanks for the suggestion :)

**Comment by glenjamin** [comment]
I guess that depends on whether the same value will be used more than once?

**Comment by jarohen** [comment]
Do you think they would be?

**Comment by glenjamin** [comment]
Possibly?

An example which springs to mind is having the client/server in an application agree on a port number or some shared secret.

**Comment by jarohen** [comment]
That's true, although if we weren't using environment variables, we'd still have to declare that twice. So probably an orthogonal concern to the environment variables.

[Nomad](https://github.com/james-henderson/nomad) had the concept of 'snippets' - pointers to another location in the config, which did solve this kind of problem - although in all my time using Nomad, I've not really used them :) Could reproduce that functionality in Phoenix if you think it'd be useful?

Without snippets, I'd probably have another common dependency (without a component) which both the server and client components could depend on - would mean an extra level of indirection when looking up the configuration value though:

``` clojure
{:common {:web-port 3000}

 :server {...
          :common ::dep}

 :client {...
          :common ::dep}}

;; looking it up:

(defrecord ServerComponent []
  c/Lifecycle
  (start [this]
    (let [web-port (get-in this [:common :web-port])]
      ...)))
```

Bit of a pain, though. What do you think?

**Comment by jarohen** [comment]
Also, worth mentioning that I'm not sure how to do client-side config support with Phoenix as yet - I don't want to pass the whole config map into CLJS because it's probably got passwords etc in that we don't want to send to the browser automatically! So far, I have (as a user of Nomad) manually passed through the specific configuration variables that the CLJS-side needs, but again, if you have any suggestions, let me know :)

**Comment by jarohen** [comment]
Hi Glen - I've just released a snapshot version - 0.0.2-SNAPSHOT - with this feature in - how does it look?

James

**Comment by glenjamin** [comment]
looks good - thanks!

**Comment by jarohen** [comment]
No problems, thanks for the suggestion :) I've just released 0.0.2 with this change in.

Cheers,

James
